src/Microsoft.Data.SqlClient/netcore/src/Microsoft/Data/SqlClientX/IO/TdsWriter.cs

```diff
@@ -176,41 +180,45 @@ public async ValueTask WriteBytesAsync(byte[] data, bool isAsync, CancellationTo
         /// <returns>Span of bytes for the TDS decimal type.</returns>
         public Span<byte> SerializeDecimal(decimal value)
```

Comment:
The function signature is incorrect for span usage. In general for low level apis you do not want to return a span you should provide one as an input parameter and then have the return value indicate if the operation succeeded. You also need an out int parameter for variable length writes, tuples are useful here. See the functions in the BCL that return OperationResult as an example of this. 

My suggested version of the SerializeDecimal function would be this:
```csharp
        public OperationStatus SerializeDecimal(decimal value, Span<byte> buffer)
        {
            if (buffer.Length < 17)
            {
                return OperationStatus.DestinationTooSmall;
            }

            Span<int> decimalBits =
#if !NET5_0_OR_GREATER
                stackalloc int[4];
            decimal.TryGetBits(value, decimalBits, out int valuesWritten);
#else
            decimal.GetBits(value);
#endif

            /*
                Returns a binary representation of a Decimal. The return value is an integer
                array with four elements. Elements 0, 1, and 2 contain the low, middle, and
                high 32 bits of the 96-bit integer part of the Decimal. Element 3 contains
                the scale factor and sign of the Decimal: bits 0-15 (the lower word) are
                unused; bits 16-23 contain a value between 0 and 28, indicating the power of
                10 to divide the 96-bit integer part by to produce the Decimal value; bits 24-
                30 are unused; and finally bit 31 indicates the sign of the Decimal value, 0
                meaning positive and 1 meaning negative.
                SQLDECIMAL/SQLNUMERIC has a byte stream of:
                struct {
                    BYTE sign; // 1 if positive, 0 if negative
                    BYTE data[]; // 16 bits = int [4]
                }
                For TDS 7.0 and above, there are always 17 bytes of data
            */
            // write the sign (note that COM and SQL are opposite)
            buffer[0] = (uint)decimalBits[3] != 0u ? (byte)1 : (byte)0;

            BinaryPrimitives.TryWriteInt32LittleEndian(buffer.Slice(1, 4), decimalBits[0]);
            BinaryPrimitives.TryWriteInt32LittleEndian(buffer.Slice(5, 4), decimalBits[1]);
            BinaryPrimitives.TryWriteInt32LittleEndian(buffer.Slice(9, 4), decimalBits[2]);
            BinaryPrimitives.TryWriteInt32LittleEndian(buffer.Slice(13, 4), 0);

            return OperationStatus.Done;
        }
```

- It uses the newest verison of TryGetBits if it can so we can avoid a heap allocation.
- It avoids using the `current` variable to keep track of the position because every offset is statically known.
- it simplifies the sign check slightly, see [sharplab](https://sharplab.io/#v2:EYLgtghglgdgNAFxBAzmAPgAQEwAYCwAUDgIxHmFHACeCApgAQCCAFAEp0QAmA8jADbUAygAcIMADywEAPgZc6AYyiR+AISgIUASiIBvIgyMNMAdgYsa9bS1ssArtO0LlqjVoDaAZgC62hgB6DLgAHgAcuJFRuP4ycl4k2gDcRAC+FFaMauycvALCYpLSci4qEOqaOvqGxmYWjjAIzkplFZ6+DACEALzB9gwA/BaZ2iQMIMO0dNq4KYTplITGDKLiUo0lLW6VDL0oCBCKANbl/AD2igzSHgAsPnPL9jCKABZKR3RcevJb5e4o3h8uwsTlCEWikSSC2W0gYEGBrFK2y0yRqRlhwGB2SRf0qqMWQA=) for the reason.
- It takes the buffer to write the decimal into as a parameter so the caller manages the renting and pooling. 

If you try return rented buffers you're going to have nasty lifetime issues with memory and you're going to spend a lot of cpu time renting returning and checking rentals etc. Don't think about it that way. Use span and memory everywhere and manage them as far up as possible.